server.py

```diff
@@ -23,35 +31,63 @@ def return_request():
     while True:
         try:
             conn, addr = socket.accept()
-            output = ""
+            request = ""
             while True:
-                msg = conn.recv(16)
-                output = output + msg
-                if len(msg) < 16:
+                msg = conn.recv(1024)
+                request = request + msg
+                if len(msg) < 1024:
                     break
             try:
-                final = parse_request(output)
-                conn.sendall(response_ok(final))
+                response = response_ok(parse_request(request))
             except ValueError as detail:
-                detail = str(detail)
-                conn.sendall(response_error("400", detail))
+                response = response_error(b"400", str(detail))
+            except LookupError as detail:
+                response = response_error(b"404", str(detail))
+            except AttributeError as detail:
+                response = response_error(b"405", str(detail))
+            except NotImplementedError as detail:
+                response = response_error(b"500", str(detail))
```

Comment:
given that you raise `NotImplementedError` for a bad http version, 500 does not seem like the right error code to return.  Shouldn't it be one of the 400-level errors?